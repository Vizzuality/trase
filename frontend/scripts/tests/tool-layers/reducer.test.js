import {
  SET_NODE_ATTRIBUTES,
  GET_CONTEXT_LAYERS,
  GET_MAP_VECTOR_DATA,
  SELECT_BASEMAP,
  SELECT_CONTEXTUAL_LAYERS,
  CHANGE_LAYOUT,
  SET_SANKEY_SIZE,
  SET_ACTIVE_MODAL,
  SELECT_UNIT_LAYERS,
  selectContextualLayers,
  TOGGLE_MAP_DIMENSION
} from 'react-components/tool/tool.actions';
import {
  saveMapView,
  TOOL_LAYERS__SAVE_MAP_VIEW,
  TOOL_LAYERS__SET_MAP_DIMENSIONS,
  TOOL_LAYERS__SET_LINKED_GEOIDS
} from 'react-components/tool-layers/tool-layers.actions';
import reducer from 'react-components/tool-layers/tool-layers.reducer';
import initialState from 'react-components/tool-layers/tool-layers.initial-state';
import { TOOL_LAYOUT, SANKEY_OFFSETS } from 'constants';

test(SET_NODE_ATTRIBUTES, () => {
  const action = {
    type: SET_NODE_ATTRIBUTES
  };
  const newState = reducer(initialState, action);
  expect(newState).toMatchSnapshot();
});

test(SET_ACTIVE_MODAL, () => {
  const action = {
    type: SET_ACTIVE_MODAL,
    payload: { activeModal: 'openedModal' }
  };
  const newState = reducer(initialState, action);
  expect(newState).toEqual({
    ...initialState,
    activeModal: 'openedModal'
  });
});

test(SELECT_UNIT_LAYERS, () => {
  const action = {
    type: SELECT_UNIT_LAYERS,
    payload: { uids: ['some_uid', 'another_uid'] }
  };
  const newState = reducer(initialState, action);
  expect(newState).toEqual({
    ...initialState,
    selectedMapDimensions: ['some_uid', 'another_uid']
  });
});

test(TOOL_LAYERS__SET_MAP_DIMENSIONS, () => {
  const dimensions = [
    {
      groupId: 3,
      layerAttributeId: 11,
      type: 'quant',
      uid: 'quant11'
    }
  ];
  const dimensionGroups = [
    { id: 3, name: 'Agricultural indicators' },
    { id: 1, name: 'Environmental indicators' }
  ];
  const action = {
    type: TOOL_LAYERS__SET_MAP_DIMENSIONS,
    payload: { dimensions, dimensionGroups }
  };
  const newState = reducer(initialState, action);
  expect(newState).toMatchSnapshot();
});

describe(TOOL_LAYERS__SET_LINKED_GEOIDS, () => {
  it('sets in linkedGeoIds an array of the nodes geoIds', () => {
    const action = {
      type: TOOL_LAYERS__SET_LINKED_GEOIDS,
      payload: { nodes: [{ id: 1, name: 'Node', geoId: 'BR-123-23' }] }
    };
    const newState = reducer(initialState, action);
    expect(newState).toMatchSnapshot();
  });

  it('sets in linkedGeoIds an empty array if there are no nodes', () => {
    const action = {
      type: TOOL_LAYERS__SET_LINKED_GEOIDS,
      payload: { nodes: [] }
    };
    const newState = reducer(initialState, action);
    expect(newState).toMatchSnapshot();
  });
});

test(GET_CONTEXT_LAYERS, () => {
  const mapContextualLayers = [
    {
      id: 10,
      identifier: 'landcover',
      isDefault: false,
      title: 'Land cover'
    }
  ];
  const action = {
    type: GET_CONTEXT_LAYERS,
    payload: { mapContextualLayers }
  };
  const newState = reducer(initialState, action);
  expect(newState).toMatchSnapshot();
});

test(GET_MAP_VECTOR_DATA, () => {
  const mapVectorData = [
    {
      id: 10,
      identifier: 'landcover',
      isDefault: false,
      title: 'Land cover'
    }
  ];
  const action = {
    type: GET_MAP_VECTOR_DATA,
    payload: { mapVectorData }
  };
  const newState = reducer(initialState, action);
  expect(newState).toMatchSnapshot();
});

test(TOOL_LAYERS__SAVE_MAP_VIEW, () => {
  const zoom = 2;
  const latlng = { lat: -34.397, lng: 150.644 };
  const action = saveMapView(latlng, zoom);

  const newState = reducer(initialState, action);
  expect(newState).toMatchSnapshot();
});

test(SELECT_BASEMAP, () => {
  const selectedBasemap = { id: 10 };
  const action = {
    type: SELECT_BASEMAP,
    payload: { selectedBasemap }
  };
  const newState = reducer(initialState, action);
  expect(newState).toMatchSnapshot();
});

test(SELECT_CONTEXTUAL_LAYERS, () => {
  const contextualLayers = [
    {
      id: 10,
      identifier: 'landcover',
      isDefault: false,
      title: 'Land cover'
    }
  ];
  const action = selectContextualLayers(contextualLayers);
  const newState = reducer(initialState, action);
  expect(newState).toMatchSnapshot();
});

describe(CHANGE_LAYOUT, () => {
  it('Changes the map layout', () => {
    const action = {
      type: CHANGE_LAYOUT,
      payload: {
        toolLayout: TOOL_LAYOUT.right
      }
    };
    const newState = reducer(initialState, action);
    expect(newState).toMatchSnapshot();
  });
});

describe(SET_SANKEY_SIZE, () => {
  it('Sets sankey size', () => {
    const action = {
      type: SET_SANKEY_SIZE
    };
    const windowInnerWidth = 1024;
    const windowInnerHeight = 768;
    const newState = reducer(initialState, action);
    expect(newState.sankeySize).toStrictEqual([
      windowInnerWidth - SANKEY_OFFSETS.splittedWidth,
      windowInnerHeight - SANKEY_OFFSETS.height
    ]);
  });
  it('Sets sankey size in map view', () => {
    const action = {
      type: SET_SANKEY_SIZE
    };
    const windowInnerWidth = 1024;
    const windowInnerHeight = 768;
    const newState = reducer({ ...initialState, toolLayout: TOOL_LAYOUT.left }, action);
    expect(newState.sankeySize).toStrictEqual([
      windowInnerWidth - SANKEY_OFFSETS.width,
      windowInnerHeight - SANKEY_OFFSETS.height
    ]);
  });
});

describe(TOGGLE_MAP_DIMENSION, () => {
  it('Remove map dimension if the dimension is selected', () => {
    const action = {
      type: TOGGLE_MAP_DIMENSION,
      payload: {
        selectedMapDimensions: ['quant95', 'quant90'],
        uid: 'quant90'
      }
    };
    const newState = reducer(initialState, action);
    expect(newState).toMatchSnapshot();
  });

  it('Add a map dimension if the dimension is not selected', () => {
    const action = {
      type: TOGGLE_MAP_DIMENSION,
      payload: {
        selectedMapDimensions: [],
        uid: 'quant90'
      }
    };
    const newState = reducer(initialState, action);
    expect(newState).toMatchSnapshot();
  });
  it('Add a map dimension in the second slot if the dimension is not selected', () => {
    const action = {
      type: TOGGLE_MAP_DIMENSION,
      payload: {
        selectedMapDimensions: ['NotEmpty'],
        uid: 'quant90'
      }
    };
    const newState = reducer(initialState, action);
    expect(newState).toMatchSnapshot();
  });
});
