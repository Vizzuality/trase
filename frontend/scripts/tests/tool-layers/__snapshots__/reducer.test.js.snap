// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CHANGE_LAYOUT Changes the map layout 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 2,
}
`;

exports[`GET_CONTEXT_LAYERS 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {
      "10": Object {
        "id": 10,
        "identifier": "landcover",
        "isDefault": false,
        "title": "Land cover",
      },
    },
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`GET_MAP_VECTOR_DATA 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": Array [
      Object {
        "id": 10,
        "identifier": "landcover",
        "isDefault": false,
        "title": "Land cover",
      },
    ],
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`SAVE_MAP_VIEW 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": Object {
    "latitude": "-34.40",
    "longitude": "150.64",
    "zoom": 2,
  },
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`SELECT_BASEMAP 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": Object {
    "id": 10,
  },
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`SELECT_CONTEXTUAL_LAYERS 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": Array [
    Object {
      "id": 10,
      "identifier": "landcover",
      "isDefault": false,
      "title": "Land cover",
    },
  ],
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`SET_NODE_ATTRIBUTES 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`TOGGLE_MAP_DIMENSION Add a map dimension if the dimension is not selected 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": true,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": Array [
    "quant90",
  ],
  "toolLayout": 1,
}
`;

exports[`TOGGLE_MAP_DIMENSION Add a map dimension in the second slot if the dimension is not selected 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": true,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": Array [
    "NotEmpty",
    "quant90",
  ],
  "toolLayout": 1,
}
`;

exports[`TOGGLE_MAP_DIMENSION Remove map dimension if the dimension is selected 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": Array [
    "quant95",
    null,
  ],
  "toolLayout": 1,
}
`;

exports[`TOOL_LAYERS__SET_LINKED_GEOIDS sets in linkedGeoIds an array of the nodes geoIds 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [
    "BR-123-23",
  ],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`TOOL_LAYERS__SET_LINKED_GEOIDS sets in linkedGeoIds an empty array if there are no nodes 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {},
    "mapDimensionsGroups": Array [],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;

exports[`TOOL_LAYERS__SET_MAP_DIMENSIONS 1`] = `
Object {
  "data": Object {
    "mapContextualLayers": Object {},
    "mapDimensions": Object {
      "quant11": Object {
        "groupId": 3,
        "layerAttributeId": 11,
        "type": "quant",
        "uid": "quant11",
      },
    },
    "mapDimensionsGroups": Array [
      Object {
        "dimensions": Array [
          "quant11",
        ],
        "id": 3,
        "name": "Agricultural indicators",
      },
      Object {
        "dimensions": Array [],
        "id": 1,
        "name": "Environmental indicators",
      },
    ],
    "mapVectorData": null,
  },
  "highlightedNodeCoordinates": null,
  "linkedGeoIds": Array [],
  "mapLoading": false,
  "mapView": null,
  "sankeySize": Array [
    1024,
    768,
  ],
  "selectedBasemap": null,
  "selectedMapContextualLayers": null,
  "selectedMapDimensions": null,
  "toolLayout": 1,
}
`;
